File: src/main/kotlin/com/leetcode/P807.kt
package com.leetcode

// https://leetcode.com/problems/max-increase-to-keep-city-skyline/
class P807 {

    fun maxIncreaseKeepingSkyline(grid: Array<IntArray>): Int {
        val size = grid.size
        val left = IntArray(size)
        val top = IntArray(size)

        for (y in 0 until size) {
            for (x in 0 until size) {
                top[x] = maxOf(top[x], grid[y][x])
                left[y] = maxOf(left[y], grid[y][x])
            }
        }

        var total = 0
        for (y in 0 until size) {
            for (x in 0 until size) {
                total += minOf(left[y], top[x]) - grid[y][x]
            }
        }
        return total
    }

}

File: src/python3/com/leetcode/p807.py
from typing import List


class Solution:
    def maxIncreaseKeepingSkyline(self, grid: List[List[int]]) -> int:
        n = len(grid)
        top = [0] * n
        left = [0] * n

        for y in range(n):
            for x in range(n):
                top[x] = max(top[x], grid[y][x])
                left[y] = max(left[y], grid[y][x])

        total = 0
        for y in range(n):
            for x in range(n):
                total += min(top[x], left[y]) - grid[y][x]

        return total
